### ISSUE TITLE
[multicluster] Fail fast caused by delayed gateway liveness probe.

### ISSUE THREAD

**Issue by chenaoxd** [linkerd/linkerd2#8590]
### What is the issue?

# Describe the bug
There will be unexpected fail-fast when access mirrored services (all the services were healthy).

# How this issue happened
1. Sometimes the client watch ([code](https://github.com/linkerd/linkerd2/blob/main/multicluster/cmd/service-mirror/main.go#L108)) will return not ok results.
    - there will be a "Link watch terminated; restarting watch"
    - but we don't know why the watch failed
2. Then the `ClusterWatcher` will be restarted ([code](https://github.com/linkerd/linkerd2/blob/main/multicluster/cmd/service-mirror/main.go#L126)).
    - the `gatewayAlive` of the new `RemoteClusterServiceWatcher` will be false
3. Although the `probeWorker` will start immediately ([code](https://github.com/linkerd/linkerd2/blob/main/multicluster/cmd/service-mirror/main.go#L190)), the ticker will trigger the first probe after `probeTickerPeriod` ([code](https://github.com/linkerd/linkerd2/blob/main/multicluster/service-mirror/jittered_ticker.go#L55)).
    - so the gatewayAliveness will be false in `probeTickerPeriod` + `maxJitter` time
4. There will be a immediate `RepairEndpoints` event ([code](https://github.com/linkerd/linkerd2/blob/main/multicluster/service-mirror/cluster_watcher.go#L881)), and then the handler will call `updateMirrorEndpoints()` ([code](https://github.com/linkerd/linkerd2/blob/main/multicluster/service-mirror/cluster_watcher.go#L1010)) in which the `updateReadiness()` will be called. And because the `gatewayAlive` may not be true yet, so all addresses will be moved to `NotReadyAdresses`.
    - then the fail-fast happened

### How can it be reproduced?

Cannot trigger manually, when the client watch ([code](https://github.com/linkerd/linkerd2/blob/main/multicluster/cmd/service-mirror/main.go#L108)) return not ok results this issue will happen.

### Logs, error output, etc

1. client watch return not ok results, the link watch terminated

    ```text
    time="2022-05-31T02:35:20Z" level=info msg="Link watch terminated; restarting watch"
    ```

2. probe starting

    ```text
    time="2022-05-31T02:35:20Z" level=info msg="Starting probe worker" probe-key=alpha
    ```

3. setting addresses to not ready

    ```text
    time="2022-05-31T02:35:21Z" level=warning msg="gateway for foo/bar does not have ready addresses; setting addresses to not ready" apiAddress="https://<kubeapi_uri>:6443" cluster=remote
    ```

4. gatewayAlive becomes true, after the "does not have ready addresses" log and almost 3s after starting the probe worker

    ```text
    time="2022-05-31T02:35:23Z" level=debug msg="gateway liveness change from false to true" apiAddress="https://39.105.197.120:6443" cluster=remote
    ```

### output of `linkerd check -o short`

```text
Linkerd extensions checks
=========================

linkerd-multicluster
--------------------
× all mirror services have endpoints
    Some mirror services do not have endpoints:
   foo-alpha.bar mirrored from cluster [alpha]
    .....
    see https://linkerd.io/2.11/checks/#l5d-multicluster-services-endpoints for hints

Status check results are ×

```

### Environment

- Kubernetes Version: v1.20.4
- Cluster Environment: ACK (Aliyun Container Service) & k8s cluster deploy by rancher rke
- Host OS: ubuntu 20.04
- Linkerd version: 2.11.2

### Possible solution

We use [this](https://github.com/chenaoxd/linkerd2/commit/d8878020d2777ec51992b8fe75c28e7762379e4f) as a hotfix , and it worked. But if the probe API is slow, this issue will still happen. Ensuring the repair happens after the correct `gatewayAlive` is set may be a good solution.

**Comment by mateiidavid** [comment]
@chenaoxd thanks for raising this! It's an interesting problem, surprised it hasn't been reported before. I want to unpack everything to make sure I got it right. First though, I think the `Link` watch is being terminated by the Kubernetes informer (perhaps due to their sync period or something?), we could add the error to the log record but not sure how valuable it would be.

For the problem itself:

* Service mirror starts a `Link` watch; watch is re-started, because of that, the `Link` **that was already present** will be added again as a new event.
* Once the watcher loop receives the event, it will do two things:

  1. Start the probe worker (which will start with a delay of probeDuration + (probeDuration / 10) which is the jitter).
  2. Re-initialize the cluster watcher and start it.

* The cluster watcher is re-initialized, because of this, `gatewayAlive` will be set to false (since we do not set it explicitly, uninitialized bools in Go are set to `false`). It will do 3 things:
   - Spawn a Go routine (or task) to handle the event loop;
   - Trigger a `RepairEndpoints` event;
   - Spawn a task for liveness.

The first `RepairEndpoints` event will resolve the gateway addresses, and then update the mirror services. Because `gatwayAlive` is false, they get marked as `NotReady`. Shortly after, the probe worker attempts to connect to the gateway, is successful, and notifies the liveness receiver that everything is `UP`. 

The liveness task will be notified and `RepairEndpoints` once again, this time **with** `gatewayAlive=true`.


---


Is my understanding accurate? What is the impact on your services? Do you get `service is in fail-fast` temporarily while all of this happens, or do your services never recover? I think regardless we want to fix this, but I'm curious how this impacts you at the moment so we can figure out what the best approach would be. If I had to guess, I'd say your calls will fail momentarily then recover, but this will happen frequently (or as frequently as the `Link` watch is terminated).

If I understand correctly, your fix is to fire the probe right away and then start the ticker. This way, when we start, the liveness task will get the opportunity to run and set `gatewayAlive=true`. But here's my question with this approach: __how does it protect us against race conditions__. Since we would issue 2 `RepairEndpoints` events, what guarantee do we have that when the first one fires, the liveness task has had the chance to be notified by the probe worker and set `gatewayAlive=true`.

In other words, my question around your fix is: are we guaranteed to **always** have `gatewayAlive=true` when the probe fires without delay? Can we guarantee that in the first `RepairEndpoints` event we process we won't mark the addresses as `NotReady`?  Perhaps we could do a bit more here to have more guarantees.

**Comment by chenaoxd** [comment]
Thank you for your reply.

> What is the impact on your services? Do you get service is in fail-fast temporarily while all of this happens, or do your services never recover? ... If I had to guess, I'd say your calls will fail momentarily then recover, but this will happen frequently (or as frequently as the Link watch is terminated).

Yes, in the 3s period, all requests to access the cross-cluster services will fail temporarily and recover after that. But even this may also be unacceptable because it will cause the whole service unavailable.

> In other words, my question around your fix is: are we guaranteed to always have gatewayAlive=true when the probe fires without delay? Can we guarantee that in the first RepairEndpoints event we process we won't mark the addresses as NotReady? Perhaps we could do a bit more here to have more guarantees.

Yes, there will be no guarantee and so I mark it as a hotfix. Maybe we can add a `Lock` to ensure `gatewayAlive` is correct before `RepairEndpoints`. But in reality, this hotfix did get rid of all the fail-fast issues we've met before. 

And I'm willing to make a more guaranteed fix when the solution is clear. 😄

**Comment by mateiidavid** [comment]
@chenaoxd sorry for being a bit slow to investigate this. The solution you proposed sounds like a good fix, but ideally we'd want to minimize any risk from race conditions.

I've spoken to some other maintainers and we think that a simple way to do this (as simple as your hotfix) would be to always instantiate the `RemoteClusterWatcher` with `gatewayAlive=true`.

We'd likely want to touch the constructor and add a new parameter, and then simply pass in a `true` bool as an argument when restarting the cluster watcher here:

https://github.com/linkerd/linkerd2/blob/716279464a2331a91f173c1b36b74a5f2f910126/multicluster/cmd/service-mirror/main.go#L197-L207

Constructor:

https://github.com/linkerd/linkerd2/blob/main/multicluster/service-mirror/cluster_watcher.go#L153-L163

We'll operate on the assumption that the gateway is alive:

1. We'll still issue the same `RepairEndpoints{}` event, which will now mark all mirrors as healthy.
2. The probe will still fire after the initial repair duration (+ jitter); if the gateway endpoints cannot be reached, they'll be marked as being down.

However, when restarting because of watch terminations, everything should work well for you since the endpoints will not be marked as being down.

WDYT about the solution? Would you be happy to submit this change?

**Comment by chenaoxd** [comment]
@mateiidavid I think this is really a good and straightforward solution. I'm willing to submit this change in one or two days.

**Comment by mateiidavid** [comment]
@chenaoxd perfect, thank you! Feel free to request me for review when it's up, and ping me if you have any questions :)

**Comment by chenaoxd** [comment]
@mateiidavid The [PR](https://github.com/linkerd/linkerd2/pull/8725) is ready. Considering it's a fairly simple change, no new tests added yet. If new tests are needed, some guidance will be helpful.
